### PR TITLE
[ADD] account_tax_update: Add a way to update taxes from local ta…

### PR DESCRIPTION
**Pull request by clbr-odoo**
…x templates

When taxes changed for a country (additions, modifications, ...) we were not applying modifications made on templates to regular tax objects. Thus users were forced to modify everything by hand or ask support to do it. We now try to safely update taxes, when we are sure they are the same we update their tax tags, else we create new ones without modifying existing ones.

Task: 3098941

### CODE REVIEW

File: addons/account_tax_update/wizard/account_tax_update.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+            return False
+        tax_rep_lines = tax.invoice_repartition_line_ids + tax.refund_repartition_line_ids
+        template_rep_lines = template.invoice_repartition_line_ids + template.refund_repartition_line_ids
+        for rep_line_tax, rep_line_template in zip(tax_rep_lines, template_rep_lines):
+            if rep_line_tax.factor_percent != rep_line_template.factor_percent:
+                return False
+        return True
+
+    def remove_xml_id(self, xml_id):
+        module, _, name = xml_id.partition(".")
```
Comment (oco-odoo) [inline addons/account_tax_update/wizard/account_tax_update.py:101]: rather use split()

File: addons/account_tax_update/wizard/account_tax_update.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+        self.env.cr.execute(
+            """
+            DELETE
+              FROM ir_model_data
+             WHERE module = %s
+               AND name = %s
+        """,
+            [module, name],
+        )
```
Comment (oco-odoo) [inline addons/account_tax_update/wizard/account_tax_update.py:110]: Rather use the ORM

File: addons/account_tax_update/wizard/account_tax_update.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        tuples = self.env.cr.fetchall()
+        template_to_tax = dict(tuples)
+        return template_to_tax
+
+    def is_same_tax_and_template_lines(self, tax, template):
+        """
+        This function compares an account.tax's and an account.tax.template's repartition lines
+        If they have a different number of repartition lines OR their factor_percent is different -> returns 'False'
+        Else -> returns 'True'
```
Comment (oco-odoo) [inline addons/account_tax_update/wizard/account_tax_update.py:88]:
```suggestion
        :return: True if ... 
```

File: addons/account_tax_update/wizard/account_tax_update.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+                    # if the tax template is not present in the mapping, it means the corresponding tax was removed from
+                    # the db, or no xmlid exists for the tax (in the case of old db or new tax)
+                    # -> create a new tax
+                    self.create_tax_from_template(company, template)
+                else:
+                    # there is a corresponding xmlid in taxes, let's check if it's exactly the same taxe
+                    tax = self.env["account.tax"].browse(template_to_tax[template.id])
+                    if tax.amount_type == template.amount_type and tax.amount == template.amount \
+                            and self.is_same_tax_and_template_lines(tax, template):
+                        # todo and tags are different else we update everytime things that are already the same
```
Comment (oco-odoo) [inline addons/account_tax_update/wizard/account_tax_update.py:29]: TODO

File: addons/account_tax_update/wizard/account_tax_update.py [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+                        # xmlid, amount_type, amount and repartition lines percentages match
+                        # -> update the tax : we only updates tax tags
+                        tax_rep_lines = tax.invoice_repartition_line_ids + tax.refund_repartition_line_ids
+                        template_rep_lines = template.invoice_repartition_line_ids + template.refund_repartition_line_ids
+                        for tax_line, template_line in zip(tax_rep_lines, template_rep_lines):
+                            tags_to_add = template_line._get_tags_to_add()
+                            # clear all tags TODO need to delete unused fields
```
Comment (oco-odoo) [inline addons/account_tax_update/wizard/account_tax_update.py:36]: todo

File: addons/account_tax_update/wizard/account_tax_update.py [hunk 6]
```diff
@@ -0,0 +1,124 @@
...
+                            tax_line.sudo().write({"tag_ids": [(5, 0, 0)]})
+                            # link new tags
+                            tax_line.sudo().write({"tag_ids": [(4, tag.id) for tag in tags_to_add]})
```
Comment (oco-odoo) [inline addons/account_tax_update/wizard/account_tax_update.py:39]: no need for the sudos here

Comment (oco-odoo) [inline addons/account_tax_update/wizard/account_tax_update.py:39]:
```suggestion
                            tax_line.sudo().write({"tag_ids": [(6, 0, tags_to_add.ids)]})
```

File: addons/account_tax_update/tests/test_account_tax_update.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+
+    def test_xmlid_match_template_match(self):
+        inv_tax_updated_tag = self.env['account.account.tag'].create({
+            'name': 'invoice_tax_updated',
+            'applicability': 'taxes',
+            'country_id': self.env.company.country_id.id,
+        })
+        self.tax.invoice_repartition_line_ids.write({'tag_ids': [(4, inv_tax_updated_tag.id, 0), (3, self.inv_tax_tag.id)]})
+        self.assertNotEqual(self.template.invoice_repartition_line_ids.tag_ids, self.tax.invoice_repartition_line_ids.tag_ids,
+                            "Tax template tags and tax tags should be different.")
```
Comment (oco-odoo) [inline addons/account_tax_update/tests/test_account_tax_update.py:63]: Probably too much; you're checking the write works ^^

File: addons/account_tax_update/tests/test_account_tax_update.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+                    (0, 0, {'factor_percent': 100, 'repartition_type': 'tax', 'tag_ids': [(4, cls.ref_tax_tag.id, 0)]}),
+                ],
+            },
+            'noupdate': True,
+        }])
+        template_vals = cls.template._get_tax_vals_complete(cls.env.company)
+        tax_id = cls.chart_template.create_record_with_xmlid(cls.env.company, cls.template, 'account.tax', template_vals)
+        cls.tax = cls.env['account.tax'].browse(tax_id)
+
+    def test_xmlid_match_template_match(self):
```
Comment (oco-odoo) [inline addons/account_tax_update/tests/test_account_tax_update.py:55]: docstring

**Review by oco-odoo — CHANGES_REQUESTED** [review]

File: addons/account_tax_update/wizard/account_tax_update.py [hunk 9]
```diff
@@ -0,0 +1,124 @@
...
+            SELECT tmplt.res_id AS tmplt_res_id,
+                   tax.res_id AS tax_res_id
+              FROM ir_model_data tax
+              JOIN ir_model_data tmplt
+                ON tmplt.name = substr(tax.name, strpos(tax.name, '_') + 1)
+             WHERE tax.model = 'account.tax'
+               AND tax.name LIKE %s
+             -- tax.name is of the form: {company_id}_{account.tax.template.name}
+            """,
+            [r"%s\_%%" % company.id],
```
[the remark this answers is not in the document]
Comment (clbr-odoo) [reply addons/account_tax_update/wizard/account_tax_update.py:78]: The first is the % for the LIKE clause and the second is the format specifier character. So yes both necessary :)

**Comment by pedrobaeza** [comment]
You may want to check https://github.com/OCA/account-financial-tools/tree/15.0/account_chart_update as inspiration :)

**Pull request closed without merging**